commcare_connect/opportunity/utils/completed_work.py

```diff
@@ -11,37 +11,74 @@
+
+    If compute_payment is True, the saved fields related to completed/approved work and payments
+    earned will also be saved against the model.
     """
     payment_accrued = 0
     for completed_work in completed_works:
-        if completed_work.completed_count < 1:
-            continue
+        payment_accrued += _update_status_set_saved_fields_and_get_payment_accrued(
+            completed_work, opportunity_access, compute_payment
+        )
 
-        if opportunity_access.opportunity.auto_approve_payments:
-            visits = completed_work.uservisit_set.values_list("status", "reason", "review_status")
-            if any(status == VisitValidationStatus.rejected for status, *_ in visits):
-                completed_work.status = CompletedWorkStatus.rejected
-                completed_work.reason = "\n".join(reason for _, reason, _ in visits if reason)
-            elif all(status == VisitValidationStatus.approved for status, *_ in visits):
-                completed_work.status = CompletedWorkStatus.approved
+    if compute_payment:
+        opportunity_access.payment_accrued = payment_accrued
+        opportunity_access.save()
 
-            if (
-                opportunity_access.opportunity.managed
-                and not all(review_status == VisitReviewStatus.agree for *_, review_status in visits)
-                and completed_work.status == CompletedWorkStatus.approved
-            ):
-                completed_work.status = CompletedWorkStatus.pending
 
-            completed_work.save()
+def _update_status_set_saved_fields_and_get_payment_accrued(completed_work, opportunity_access, compute_payment):
+    completed_count = completed_work.completed_count
+    if completed_count < 1:
+        return 0
 
-        if compute_payment:
-            approved_count = completed_work.approved_count
-            if approved_count > 0 and completed_work.status == CompletedWorkStatus.approved:
-                payment_accrued += approved_count * completed_work.payment_unit.amount
+    amount_accrued = 0
+    made_changes = False
+    if opportunity_access.opportunity.auto_approve_payments:
+        visits = completed_work.uservisit_set.values_list("status", "reason", "review_status")
+        if any(status == VisitValidationStatus.rejected for status, *_ in visits):
+            completed_work.status = CompletedWorkStatus.rejected
+            completed_work.reason = "\n".join(reason for _, reason, _ in visits if reason)
+        elif all(status == VisitValidationStatus.approved for status, *_ in visits):
+            completed_work.status = CompletedWorkStatus.approved
+
+        if (
+            opportunity_access.opportunity.managed
+            and not all(review_status == VisitReviewStatus.agree for *_, review_status in visits)
+            and completed_work.status == CompletedWorkStatus.approved
+        ):
+            completed_work.status = CompletedWorkStatus.pending
+
+        made_changes = True
```

Comment:
nit: this is only true if  one of the above `if`s is true, right? In this implementation `made_changes == opportunity_access.opportunity.auto_approve_payments` rather than any actual changes?

Reply:
Yeah that's right, which I did intentionally since that's also [how the previous version's save logic worked](https://github.com/dimagi/commcare-connect/blob/main/commcare_connect/opportunity/utils/completed_work.py#L20-L35). We could theoretically make it more specific, but I was worried about accidentally changing behavior.